app/uk/gov/hmrc/securemessage/repository/ConversationRepository.scala

```diff
@@ -54,13 +60,15 @@ class ConversationRepository @Inject()(implicit connector: MongoConnector)
         unique = true,
         sparse = true))
 
-  def insertIfUnique(conversation: Conversation)(implicit ec: ExecutionContext): Future[Boolean] =
+  def insertIfUnique(conversation: Conversation)(
+    implicit ec: ExecutionContext): Future[Either[StoreException, Boolean]] =
     insert(conversation)
-      .map(_ => true)
+      .map(_ => Right(true))
       .recoverWith {
         case e: DatabaseException if e.code.contains(DuplicateKey) =>
-          logger.warn("Ignoring duplicate found on insertion to conversation collection: " + e.getMessage())
-          Future.successful(false)
+          val errMsg = "Ignoring duplicate found on insertion to conversation collection: " + e.getMessage()
+          logger.error(errMsg, e)
+          Future.successful(Left(StoreException(CONFLICT, errMsg, Some(e))))
```

Comment:
can we have a specialised exeption: DuplicateConversationException(message, error) extends StoreException. 
trait StoreException {
    def message: String
    def exception: Throwable
}

Reply:
StoreException was the one you created for this scenario

Reply:
StoreException is a case class not a trait

Reply:
do we pass a throwable or root exception here too or just a msg?

Reply:
yes, and now i realise we need a better/more specialised exception :D

Reply:
> do we pass a throwable or root exception here too or just a msg?

yes we should pass the parent exception, the type should be whatever is in java.lang.Exception, as that should be our base.

Reply:
> StoreException was the one you created for this scenario

yes, and now i realise we need a more specialised exception :)

Reply:
> StoreException is a case class not a trait

yes, and i suggest we have the types and  Hierarchy like like this: 
class SecureMessageException(message: String, cause: Throwable) extends Exception(message, cause) this should be the top level exception
class StoreException(message: String, cause: Throwable) extends SecureMessageException(message, cause) - top level for all DB exceptions
case class DuplicateConversationException(message: String, cause: Throwable) extends StoreException(message, cause) - specialised exception for duplicate conversation at db level.
Same pattern followed for other parts, ex: 
class ConnectorException(message: String, cause: Throwable) extends SecureMessageException(message, cause)
case class EmailLookupException(message: String, cause: Throwable) extends ConnectorException(message, cause)

Bottom line we should not have code in any of the exceptions, we should have specialised exeptions and map them to the HTTP codes in the controllers only. If an upsteam system responds with a non success http code, then our code will be 502 and the message will say: couldn't retrieve: what_data from upstream system x, return error code: xxx, with the error message: xxx.

Reply:
So you are saying now we need to create an instance of Throwable in every class.

Reply:
you are right, removed extending from Exception, made cause option, and replace Exception suffix with Error.